File: macos/Sources/ServerlifeCore/Supervisor.swift
import Darwin
import Foundation

public enum ManagedState: Sendable { case stopped, starting, running, restarting, failed }

/// A server Serverlife is responsible for: one we started, or an external one that was
/// adopted. Holds everything needed to bring it back after it dies.
public final class ManagedServer: @unchecked Sendable {
    public var name: String
    public var directory: String
    /// Shell command to run. Empty means serve `directory` with the built-in server.
    public var command: String
    public var autoRestart: Bool = true
    public var state: ManagedState = .stopped
    /// Learned on first successful start, then used as the health check.
    public var port: Int?
    public var restartCount: Int = 0
    public var lastError: String?

    var processGroup: ProcessGroup?
    var builtIn: StaticServer?
    var healthySinceUtc: Date?
    var retryAfterUtc: Date?
    var consecutiveFailures: Int = 0

    /// Last 500 output lines, for the row's log drawer.
    private let logLock = NSLock()
    private var log: [String] = []

    public init(name: String, directory: String, command: String, autoRestart: Bool = true) {
        self.name = name
        self.directory = directory
        self.command = command
        self.autoRestart = autoRestart
    }

    public var usesBuiltInServer: Bool { command.trimmingCharacters(in: .whitespaces).isEmpty }

    public func snapshot() -> [String] {
        logLock.lock(); defer { logLock.unlock() }
        return log
    }

    func append(_ line: String) {
        logLock.lock()
        log.append(line)
        if log.count > 500 { log.removeFirst(log.count - 500) }
        logLock.unlock()
    }
}

/// Starts, stops and keeps alive the servers Serverlife manages.
///
/// The watchdog is the reason the app exists: a dev server that dies mid-presentation
/// should be back before anyone notices, without alt-tabbing to a terminal.
public final class Supervisor: @unchecked Sendable {
    private static let watchdogInterval: TimeInterval = 3

    /// Backoff between restart attempts. A server that fails instantly and repeatedly is
    /// usually broken rather than unlucky, and hammering it just burns CPU during a talk.
    private static let backoff: [TimeInterval] = [1, 2, 4, 8, 15]
    private static let maxConsecutiveFailures = 10

    /// How long a server must stay up before its failure streak is forgiven. Without
    /// this, something that dies once an hour would exhaust its ten attempts over a day
    /// and then stay down — the exact opposite of the point.
    private static let healthyResetAfter: TimeInterval = 60

    private let lock = NSLock()
    private var _servers: [ManagedServer] = []
    private var watchdogTask: Task<Void, Never>?

    public init() {}

    public var servers: [ManagedServer] {
        lock.lock(); defer { lock.unlock() }
        return _servers
    }

    /// Raised whenever a managed server changes state, on a background thread.
    public var onChanged: (@Sendable (ManagedServer) -> Void)?

    public func start() {
        guard watchdogTask == nil else { return }
        watchdogTask = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: UInt64(Self.watchdogInterval * 1_000_000_000))
                self?.tick()
            }
        }
    }

    @discardableResult
    public func add(_ server: ManagedServer) -> ManagedServer {
        lock.lock(); _servers.append(server); lock.unlock()
        return server
    }

    public func remove(_ server: ManagedServer) {
        stop(server)
        lock.lock(); _servers.removeAll { $0 === server }; lock.unlock()
    }

    /// Takes over an already-running server so the watchdog can bring it back if it
    /// dies. It keeps running as it is — adopting does not restart it.
    @discardableResult
    public func adopt(_ detected: DetectedServer) -> ManagedServer {
        let server = ManagedServer(
            name: detected.displayName, directory: detected.workingDirectory ?? "",
            command: detected.commandLine ?? "")
        server.port = detected.port
        server.state = .running
        server.healthySinceUtc = Date()
        server.append("[serverlife] adopted PID \(detected.pid) on port \(detected.port)")
        return add(server)
    }

    // ---- start / stop --------------------------------------------------------------

    public func startServer(_ server: ManagedServer) {
        stop(server)

        server.lastError = nil
        server.state = .starting
        onChanged?(server)

        do {
            if server.usesBuiltInServer {
                try startBuiltIn(server)
            } else {
                try startProcess(server)
            }
            server.state = .running
            server.healthySinceUtc = Date()
        } catch {
            server.state = .failed
            server.lastError = error.localizedDescription
            server.append("[serverlife] start failed: \(error.localizedDescription)")
        }

        onChanged?(server)
    }

    private func startBuiltIn(_ server: ManagedServer) throws {
        let built = try StaticServer(folder: server.directory)
        built.start()
        server.builtIn = built
        server.port = built.port
        server.append("[serverlife] built-in static server on http://localhost:\(built.port)/")
    }

    private func startProcess(_ server: ManagedServer) throws {
        let group = ProcessGroup()
        try group.spawn(command: server.command, workingDirectory: server.directory) { [weak server] line in
            server?.append(line)
        }
        server.processGroup = group
        server.append("[serverlife] started: \(server.command)")
    }

    public func stop(_ server: ManagedServer) {
        // Killing the process group takes the whole tree: "npm run dev" is
        // sh -> npm -> node, and killing only the process we launched would leave node
        // holding the port.
        server.processGroup?.kill()
        server.processGroup = nil

        server.builtIn?.stop()
        server.builtIn = nil

        server.healthySinceUtc = nil
        if server.state != .failed {
            server.state = .stopped
        }
        onChanged?(server)
    }

    // ---- watchdog --------------------------------------------------------------

    private func tick() {
        let listening = PortScanner.getListeners()

        for server in servers {
            if server.state == .stopped || server.state == .starting { continue }

            if isHealthy(server, listening: listening) {
                onHealthy(server)
                continue
            }

            if !server.autoRestart {
                if server.state != .stopped {
                    server.state = .stopped
                    onChanged?(server)
                }
                continue
            }

            attemptRestart(server)
        }
    }

    /// Healthy means both that the process is alive and that its port is still
    /// accepting. The port matters on its own: a dev server can wedge, keep its process,
    /// and stop listening — from the outside that is indistinguishable from a crash.
    private func isHealthy(_ server: ManagedServer, listening: [Listener]) -> Bool {
        if server.usesBuiltInServer {
            return server.builtIn != nil
        }

        guard let group = server.processGroup, let pid = group.pid else { return false }
        guard kill(pid, 0) == 0 else { return false }

        guard let port = server.port else {
            // Not yet known: still starting. Learn it from whichever process in our
            // group opened a socket — with "npm run dev" that is a grandchild we never
            // held a handle to, so the group is the only reliable way to recognise it.
            if let found = listening.first(where: { group.contains($0.pid) }) {
                server.port = found.port
                server.append("[serverlife] listening on http://localhost:\(found.port)/")
            }
            // An adopted or just-started server without a port yet is given the benefit
            // of the doubt rather than being restarted out from under itself.
            return true
        }

        return listening.contains { $0.port == port }
    }

    private func onHealthy(_ server: ManagedServer) {
        if server.healthySinceUtc == nil { server.healthySinceUtc = Date() }

        if server.consecutiveFailures > 0,
           let since = server.healthySinceUtc, Date().timeIntervalSince(since) >= Self.healthyResetAfter {
            server.consecutiveFailures = 0
            server.retryAfterUtc = nil
        }

        if server.state != .running {
            server.state = .running
            onChanged?(server)
        }
    }

    private func attemptRestart(_ server: ManagedServer) {
        let now = Date()
        if let after = server.retryAfterUtc, now < after { return }

        if server.consecutiveFailures >= Self.maxConsecutiveFailures {
            if server.state != .failed {
                server.state = .failed
                server.lastError = "Gave up after \(Self.maxConsecutiveFailures) restart attempts."
                server.append("[serverlife] \(server.lastError!)")
                onChanged?(server)
            }
            return
        }

        server.state = .restarting
        server.consecutiveFailures += 1
        server.restartCount += 1
        server.healthySinceUtc = nil
        onChanged?(server)

        server.append("[serverlife] died; restart attempt \(server.consecutiveFailures)")
        startServer(server)

        // Schedule the next attempt from the end of this one, so a server that fails
        // instantly still respects the backoff.
        let wait = Self.backoff[min(server.consecutiveFailures - 1, Self.backoff.count - 1)]
        server.retryAfterUtc = Date().addingTimeInterval(wait)
    }

    public func dispose() {
        watchdogTask?.cancel()
        watchdogTask = nil
        for server in servers { stop(server) }
    }
}

File: src/Serverlife/Core/Supervisor.cs
using System.Diagnostics;
using System.IO;

namespace Serverlife.Core;

public enum ManagedState { Stopped, Starting, Running, Restarting, Failed }

/// <summary>
/// A server Serverlife is responsible for: one we started, or an external one that was
/// adopted. Holds everything needed to bring it back after it dies.
/// </summary>
public sealed class ManagedServer
{
    public required string Name { get; set; }
    public required string Directory { get; set; }

    /// <summary>Shell command to run. Empty means serve <see cref="Directory"/> with the built-in server.</summary>
    public required string Command { get; set; }

    public bool AutoRestart { get; set; } = true;
    public ManagedState State { get; set; } = ManagedState.Stopped;

    /// <summary>Learned on first successful start, then used as the health check.</summary>
    public int? Port { get; set; }

    public int RestartCount { get; set; }
    public string? LastError { get; set; }

    internal Process? Process { get; set; }
    internal JobObject? Job { get; set; }
    internal StaticServer? BuiltIn { get; set; }
    internal DateTime? HealthySinceUtc { get; set; }
    internal DateTime? RetryAfterUtc { get; set; }
    internal int ConsecutiveFailures { get; set; }

    /// <summary>Last 500 output lines, for the row's log drawer.</summary>
    internal readonly Queue<string> Log = new();

    public bool UsesBuiltInServer => string.IsNullOrWhiteSpace(Command);

    public IReadOnlyList<string> Snapshot()
    {
        lock (Log)
            return Log.ToList();
    }

    internal void Append(string line)
    {
        lock (Log)
        {
            Log.Enqueue(line);
            while (Log.Count > 500)
                Log.Dequeue();
        }
    }
}

/// <summary>
/// Starts, stops and keeps alive the servers Serverlife manages.
///
/// The watchdog is the reason the app exists: a dev server that dies mid-presentation
/// should be back before the audience notices, without anyone alt-tabbing to a terminal.
/// </summary>
public sealed class Supervisor : IDisposable
{
    private static readonly TimeSpan WatchdogInterval = TimeSpan.FromSeconds(3);

    /// <summary>
    /// Backoff between restart attempts. A server that fails instantly and repeatedly is
    /// usually broken rather than unlucky, and hammering it just burns CPU during a talk.
    /// </summary>
    private static readonly TimeSpan[] Backoff =
    [
        TimeSpan.FromSeconds(1),
        TimeSpan.FromSeconds(2),
        TimeSpan.FromSeconds(4),
        TimeSpan.FromSeconds(8),
        TimeSpan.FromSeconds(15),
    ];

    private const int MaxConsecutiveFailures = 10;

    /// <summary>
    /// How long a server must stay up before its failure streak is forgiven. Without
    /// this, something that dies once an hour would exhaust its ten attempts over a day
    /// and then stay down — the exact opposite of the point.
    /// </summary>
    private static readonly TimeSpan HealthyResetAfter = TimeSpan.FromSeconds(60);

    private readonly List<ManagedServer> _servers = new();
    private readonly CancellationTokenSource _cts = new();

    public IReadOnlyList<ManagedServer> Servers
    {
        get { lock (_servers) return _servers.ToList(); }
    }

    /// <summary>Raised whenever a managed server changes state, on a background thread.</summary>
    public event Action<ManagedServer>? Changed;

    public void Start() => _ = WatchdogLoopAsync(_cts.Token);

    public ManagedServer Add(ManagedServer server)
    {
        lock (_servers)
            _servers.Add(server);
        return server;
    }

    public void Remove(ManagedServer server)
    {
        Stop(server);
        lock (_servers)
            _servers.Remove(server);
    }

    /// <summary>
    /// Takes over an already-running server so the watchdog can bring it back if it dies.
    /// It keeps running as it is — adopting does not restart it.
    /// </summary>
    public ManagedServer Adopt(DetectedServer detected)
    {
        var server = new ManagedServer
        {
            Name = detected.DisplayName,
            Directory = detected.WorkingDirectory ?? "",
            Command = detected.CommandLine ?? "",
            Port = detected.Port,
            State = ManagedState.Running,
            HealthySinceUtc = DateTime.UtcNow,
        };
        server.Append($"[serverlife] adopted PID {detected.Pid} on port {detected.Port}");
        return Add(server);
    }

    // ---- start / stop ------------------------------------------------------------

    public void StartServer(ManagedServer server)
    {
        Stop(server);

        server.LastError = null;
        server.State = ManagedState.Starting;
        Changed?.Invoke(server);

        try
        {
            if (server.UsesBuiltInServer)
                StartBuiltIn(server);
            else
                StartProcess(server);

            server.State = ManagedState.Running;
            server.HealthySinceUtc = DateTime.UtcNow;
        }
        catch (Exception e)
        {
            server.State = ManagedState.Failed;
            server.LastError = e.Message;
            server.Append($"[serverlife] start failed: {e.Message}");
        }

        Changed?.Invoke(server);
    }

    private static void StartBuiltIn(ManagedServer server)
    {
        var built = new StaticServer(server.Directory);
        built.Start();
        server.BuiltIn = built;
        server.Port = built.Port;
        server.Append($"[serverlife] built-in static server on http://localhost:{built.Port}/");
    }

    private void StartProcess(ManagedServer server)
    {
        var job = new JobObject();

        var info = new ProcessStartInfo
        {
            FileName = "cmd.exe",
            // /C so cmd exits once the command does. The command is passed as one quoted
            // unit so its own quotes survive cmd's parsing.
            Arguments = $"/C \"{server.Command}\"",
            WorkingDirectory = server.Directory,
            UseShellExecute = false,
            // Without this every start flashes a console window on screen, which during a
            // presentation is exactly the thing this app is meant to prevent.
            CreateNoWindow = true,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
        };

        // A GUI app keeps the PATH it was launched with, so a Node that was installed (or
        // a version manager that re-pointed) since the last sign-in isn't on it and
        // "npm run dev" dies with "is not recognized". Hand the child the PATH a fresh
        // logon would have instead. info.Environment is pre-seeded with our own block, so
        // this replaces just the one entry.
        info.Environment["Path"] = EnvironmentPath.Value;

        var process = Process.Start(info)
            ?? throw new InvalidOperationException("Process.Start returned no process.");

        // Assign before the child has a chance to spawn anything, so descendants are
        // covered by the job from the outset.
        job.Assign(process);

        process.OutputDataReceived += (_, e) => { if (e.Data is not null) server.Append(e.Data); };
        process.ErrorDataReceived += (_, e) => { if (e.Data is not null) server.Append(e.Data); };
        process.BeginOutputReadLine();
        process.BeginErrorReadLine();

        server.Process = process;
        server.Job = job;
        server.Append($"[serverlife] started: {server.Command}");
    }

    public void Stop(ManagedServer server)
    {
        // Killing the job takes the whole tree: "npm run dev" is cmd -> npm -> node, and
        // killing only the process we launched would leave node holding the port.
        server.Job?.Kill();
        server.Job?.Dispose();
        server.Job = null;

        try
        {
            server.Process?.Dispose();
        }
        catch (InvalidOperationException)
        {
        }
        server.Process = null;

        server.BuiltIn?.Dispose();
        server.BuiltIn = null;

        server.HealthySinceUtc = null;
        if (server.State is not ManagedState.Failed)
            server.State = ManagedState.Stopped;
        Changed?.Invoke(server);
    }

    // ---- watchdog ----------------------------------------------------------------

    private async Task WatchdogLoopAsync(CancellationToken token)
    {
        using var timer = new PeriodicTimer(WatchdogInterval);
        while (await timer.WaitForNextTickAsync(token).ConfigureAwait(false))
        {
            try
            {
                Tick();
            }
            catch (Exception e)
            {
                Debug.WriteLine($"[Serverlife] watchdog tick failed: {e}");
            }
        }
    }

    private void Tick()
    {
        var listening = PortScanner.GetListeners();

        foreach (var server in Servers)
        {
            if (server.State is ManagedState.Stopped or ManagedState.Starting)
                continue;

            if (IsHealthy(server, listening))
            {
                OnHealthy(server);
                continue;
            }

            if (!server.AutoRestart)
            {
                if (server.State is not ManagedState.Stopped)
                {
                    server.State = ManagedState.Stopped;
                    Changed?.Invoke(server);
                }
                continue;
            }

            AttemptRestart(server);
        }
    }

    /// <summary>
    /// Healthy means both that the process is alive and that its port is still accepting.
    /// The port matters on its own: a dev server can wedge, keep its process, and stop
    /// listening — from the audience's side that is indistinguishable from a crash.
    /// </summary>
    private static bool IsHealthy(ManagedServer server, List<Listener> listening)
    {
        if (server.UsesBuiltInServer)
            return server.BuiltIn is not null;

        if (server.Process is { HasExited: true })
            return false;

        if (server.Port is not { } port)
        {
            // Not yet known: still starting. Learn it from whichever process in our job
            // opened a socket - with "npm run dev" that is a grandchild we never held a
            // handle to, so the job is the only reliable way to recognise it as ours.
            if (server.Job is { } job)
            {
                var found = listening.FirstOrDefault(l => job.Contains(l.Pid));
                if (found.Port != 0)
                {
                    server.Port = found.Port;
                    server.Append($"[serverlife] listening on http://localhost:{found.Port}/");
                }
            }
            // An adopted or just-started server without a port yet is given the benefit
            // of the doubt rather than being restarted out from under itself.
            return true;
        }

        return listening.Any(l => l.Port == port);
    }

    private void OnHealthy(ManagedServer server)
    {
        server.HealthySinceUtc ??= DateTime.UtcNow;

        if (server.ConsecutiveFailures > 0
            && DateTime.UtcNow - server.HealthySinceUtc >= HealthyResetAfter)
        {
            server.ConsecutiveFailures = 0;
            server.RetryAfterUtc = null;
        }

        if (server.State is not ManagedState.Running)
        {
            server.State = ManagedState.Running;
            Changed?.Invoke(server);
        }
    }

    private void AttemptRestart(ManagedServer server)
    {
        var now = DateTime.UtcNow;
        if (server.RetryAfterUtc is { } after && now < after)
            return;

        if (server.ConsecutiveFailures >= MaxConsecutiveFailures)
        {
            if (server.State is not ManagedState.Failed)
            {
                server.State = ManagedState.Failed;
                server.LastError = $"Gave up after {MaxConsecutiveFailures} restart attempts.";
                server.Append($"[serverlife] {server.LastError}");
                Changed?.Invoke(server);
            }
            return;
        }

        server.State = ManagedState.Restarting;
        server.ConsecutiveFailures++;
        server.RestartCount++;
        server.HealthySinceUtc = null;
        Changed?.Invoke(server);

        server.Append($"[serverlife] died; restart attempt {server.ConsecutiveFailures}");
        StartServer(server);

        // Schedule the next attempt from the end of this one, so a server that fails
        // instantly still respects the backoff.
        var wait = Backoff[Math.Min(server.ConsecutiveFailures - 1, Backoff.Length - 1)];
        server.RetryAfterUtc = DateTime.UtcNow + wait;
    }

    public void Dispose()
    {
        _cts.Cancel();
        foreach (var server in Servers)
            Stop(server);
        _cts.Dispose();
    }
}
